edit/edit_test.go

```diff
@@ -216,3 +217,57 @@ rule()`, 1, true, false, `Error for multiple unnamed rules`},
 		}
 	}
 }
+
+func TestListSubstitute(t *testing.T) {
+	tests := []struct {
+		desc, input, oldPattern, newTemplate, expected string
+	}{
+		{
+			desc:        "no_match",
+			input:       `["abc"]`,
+			oldPattern:  `!!`,
+			newTemplate: `xx`,
+			expected:    `["abc"]`,
+		}, {
+			desc:        "full_match",
+			input:       `["abc"]`,
+			oldPattern:  `.*`,
+			newTemplate: `xx`,
+			expected:    `["xx"]`,
+		}, {
+			desc:        "partial_match",
+			input:       `["abcde"]`,
+			oldPattern:  `bcd`,
+			newTemplate: `xyz`,
+			expected:    `["axyze"]`,
+		}, {
+			desc:        "number_group",
+			input:       `["abcde"]`,
+			oldPattern:  `a(bcd)`,
+			newTemplate: `$1 $1`,
+			expected:    `["bcd bcde"]`,
+		}, {
+			desc:        "name_group",
+			input:       `["abcde"]`,
+			oldPattern:  `a(?P<x>bcd)`,
+			newTemplate: `$x $x`,
+			expected:    `["bcd bcde"]`,
+		},
+	}
+
+	for _, tst := range tests {
+		t.Run(tst.desc, func(t *testing.T) {
+			f, err := build.Parse("BUILD", []byte(tst.input))
+			if err != nil {
+				t.Fatal(err)
+			}
+			lst := f.Stmt[0]
+			oldRegexp := regexp.MustCompile(tst.oldPattern)
+			ListSubstitute(lst, oldRegexp, tst.newTemplate)
+			got := build.FormatString(lst)
+			if got != tst.expected {
+				t.Fatalf("got %s ; want %s", got, tst.expected)
```

Comment:
Tell the user what happened (and prefer Errorf over Fatalf)

t.Errorf("build.FormatString(%v) = %q; want %q", lst, got, tst.expected)

Reply:
Done. I used `ListSubstitute` instead of `FormatString` in the new message to make it clear that's the operation that's failing. The input to `FormatString` won't be readable.